File: Mue.Clients.Common/ClientServer/HubClientSession.cs
using System;
using System.Threading.Tasks;
using Microsoft.AspNetCore.SignalR.Client;
using Microsoft.Extensions.DependencyInjection;
using Mue.Common.ClientServer;

namespace Mue.Clients.ClientServer
{
    public class HubClientSession : IAsyncDisposable
    {
        private readonly string _url;
        private readonly IServerCommands _commandHandler;
        private HubConnection? _connection;

        public HubClientSession(string url, IServerCommands commandHandler)
        {
            _url = url;
            _commandHandler = commandHandler;
        }

        public HubClient? Client { get; private set; }
        public bool IsConnected { get { return _connection != null && _connection.State != HubConnectionState.Disconnected; } }

        public async Task Open()
        {
            await OpenHubConnection();
        }

        public async Task Close()
        {
            if (_connection != null)
            {
                await _connection.StopAsync();
            }
        }

        public async ValueTask DisposeAsync()
        {
            await Close();
            Client?.Dispose();
        }

        private async Task<HubConnection> OpenHubConnection()
        {
            _connection = new HubConnectionBuilder()
                .WithUrl(_url)
                .WithAutomaticReconnect()
                .AddNewtonsoftJsonProtocol()
                .Build();

            // TODO: Handle other state changes like reconnecting (notify connected users?)
            _connection.Closed += OnHubClosed;

            Client = new HubClient(_connection, _commandHandler);

            await _connection.StartAsync();

            return _connection;
        }

        private void OnHubError(Exception e)
        {
            // TODO: Notify consumer of hub error
            Console.WriteLine("Hub error: " + e.ToString());
        }

        private Task OnHubClosed(Exception? e)
        {
            // TODO: Handle hub closure
            Console.WriteLine("Hub closed! " + e?.Message);
            return Task.CompletedTask;
        }
    }
}

File: Mue.Clients.Web/src/app/mue/HubClientSession.ts
import * as signalR from "@microsoft/signalr";
import { HubClient, HubClientCallbacks } from "./HubClient";

export class HubClientSession {
    private _connection?: signalR.HubConnection;
    private _client?: HubClient;

    constructor(
        private url: string,
        private callbacks: HubClientCallbacks,
    ) {}

    get client() {
        return this._client;
    }

    get isAuthenticated() {
        return this._client?.isAuthenticated || false;
    }

    async open() {
        this._connection = new signalR.HubConnectionBuilder()
            .withUrl(this.url)
            .withAutomaticReconnect()
            .build();

        this._connection.onclose((err) => {
            this._client = undefined;
            console.log("Connection closed by server", err);
        });

        await this._connection.start();

        this._client = new HubClient(this._connection, this.callbacks);
    }

    async close() {
        if (this._connection) {
            await this._connection.stop();
            console.log("Connection closed by user");
        }
    }
}
